### PR TITLE
feat(plugins) anonymous authentication in auth plugins

### LINKED ISSUE #853: allow for anonymous access to APIs with authentication
an "anonymous" consumer can be used to refer to none-authenticated caller to the API, perhaps an internal reference or an operator created one with a flag `anonymous: true`

should also allow for setting plugin rules as regular consumers, used when authentication plugins are present and no Authentication headers are sent.

### PR DESCRIPTION
**Pull request by thefosk**
### Summary

Adds support for anonymous authentication in every authentication plugin. When `config.anonymous=true`, even non-authenticated requests are allowed and the plugins will append a `X-Anonymous-Consumer: true` header to the upstream request.
### Full changelog
- basic-auth: support for `anonymous` consumers
- key-auth: support for `anonymous` consumers
- oauth2: support for `anonymous` consumers
- hmac-auth: support for `anonymous` consumers
- jwt: support for `anonymous` consumers
- ldap-auth: support for `anonymous` consumers
### Issues resolved

Fix #853.

### CODE REVIEW

**Comment by Tieske** [comment]
wouldn't this automatically add multi-auth support (in a logical OR fashion?)

**Comment by thefosk** [comment]
@Tieske yes, if all of them have `config.anonymous=true` (which means ultimately the consumer can make a request without specifying any authentication). No, if you want at least one authentication to authenticate the consumer.

File: kong/plugins/basic-auth/access.lua [hunk 1]
```diff
@@ -101,7 +105,7 @@ function _M.execute(conf)
   end
 
   if not credential or not validate_credentials(credential, given_password) then
-    return responses.send_HTTP_FORBIDDEN("Invalid authentication credentials")
+    return false, {status = 403, message = "Invalid authentication credentials"}
```
Comment (Tieske) [inline kong/plugins/basic-auth/access.lua]: why not just return `success`, `statuscode`, `message` without creating the intermediate table? More efficient and less GC.

same for the other plugins

File: kong/plugins/basic-auth/access.lua [hunk 2]
```diff
@@ -113,12 +117,26 @@ function _M.execute(conf)
...
 
-  ngx.req.set_header(constants.HEADERS.CONSUMER_ID, consumer.id)
-  ngx.req.set_header(constants.HEADERS.CONSUMER_CUSTOM_ID, consumer.custom_id)
-  ngx.req.set_header(constants.HEADERS.CONSUMER_USERNAME, consumer.username)
-  ngx.req.set_header(constants.HEADERS.CREDENTIAL_USERNAME, credential.username)
+  ngx_set_header(constants.HEADERS.ANONYMOUS, nil) -- In case of auth plugins concatenation
+  ngx_set_header(constants.HEADERS.CONSUMER_ID, consumer.id)
+  ngx_set_header(constants.HEADERS.CONSUMER_CUSTOM_ID, consumer.custom_id)
+  ngx_set_header(constants.HEADERS.CONSUMER_USERNAME, consumer.username)
+  ngx_set_header(constants.HEADERS.CREDENTIAL_USERNAME, credential.username)
```
Comment (Tieske) [inline kong/plugins/basic-auth/access.lua]: can't we cache those constants locally? the above are 15 table lookups

same for the other plugins

File: kong/constants.lua [hunk 3]
```diff
@@ -26,7 +26,8 @@ return {
     RATELIMIT_REMAINING = "X-RateLimit-Remaining",
     CONSUMER_GROUPS = "X-Consumer-Groups",
     FORWARDED_HOST = "X-Forwarded-Host",
-    FORWARDED_PREFIX = "X-Forwarded-Prefix"
+    FORWARDED_PREFIX = "X-Forwarded-Prefix",
+    ANONYMOUS = "X-Anonymous-Consumer"
```
Comment (Tieske) [inline kong/constants.lua]: please use trailing comma for table constructors, to keep the diff clean

same for the other plugins

**Comment by Suraj-Kumar** [comment]
+1

**Comment by rajskhanna** [comment]
Thanks Tieske. So it looks like if I want to force one authentication or the other it is not possible. Excluding the anonymous access. So as I see it we should be defining 2 paths to accomplish this. right?

**Pull request merged**